### PR TITLE
Custom Rust Toolchain for AVR

### PR DESCRIPTION
**Pull request by stappersg**
Implements a compiler bug fix proposal.
Un-reverts the llvm_asm! removal.

To get a workable Rust for AVR microcontrollers.

### CODE REVIEW

**Comment by nellshamrell** [comment]
Hi @stappersg! Thank you for your submission to This Week in Rust.

Unfortunately, this reddit post does not contain enough content or context to be included in the newsletter. Would you be willing to write up a more detailed post including things like why you created this custom toolchain and what your learned in the process.

Closing this for now.

**Comment by drmorr0** [comment]
@nellshamrell Hi, just to provide a little more context, the reddit post in question is mine.  The tl;dr is that if you are trying to use Rust with AVR, you are (currently) stuck on a version of rustc that is almost a year and a half old at this point.  I have a custom build of both LLVM and rustc that is based off a more modern "nightly" version of rust that includes some of the really nice features that have landed in the last year and a half -- e.g., faster compile times, some AVR-specific bugfixes, etc.

I'm not totally opposed to writing up a blog post about some of the things I've been working on; I have a (personal) blog at https://objectdisoriented.evokewonder.com, though it's rather outdated at this point.  But I'd be curious to know more about this process/what you would be hoping to get out of it :)

**Comment by stappersg** [comment]
On Fri, Apr 08, 2022 at 08:19:49PM -0700, David wrote:
>  ...  But I'd be curious to know more about this process/what you would be hoping to get out of it :)

What I hope for getting with this merge request is
more attention for the current situation
about Rust for embedded systems with an AVR.


Groeten
Geert Stappers
-- 
Silence is hard to parse

**Comment by ericseppanen** [comment]
As a reader, I would love to hear more about the history and future of Rust on AVR: where did this old Rust AVR version come from, and why is it not maintained? Is there a plan for this new fork to be maintained? What would it take for AVR support to be part of mainline Rust? I'm not asking anyone to point fingers; just trying to understand how it came to be this way and if things might get better in the future.

**Comment by ericseppanen** [comment]
Oops, didn't mean to re-open, sorry. Re-closing.

**Comment by stappersg** [comment]
On Sun, Apr 10, 2022 at 12:31:27PM -0700, Eric Seppanen wrote:
> Oops, didn't mean to re-open, sorry. Re-closing.

For the "tell me more": https://github.com/rust-lang/compiler-builtins/issues/400

**Comment by ericseppanen** [comment]
I'm not asking for those answers here; I'm suggesting that this is the kind of context that would make for a good blog post.

**Comment by timwalls** [comment]
I wrote a blog post on this a few months ago when I cobbled together my own AVR-rust toolchain (https://snowgoons.ro/posts/2021-10-09-the-day-rust-on-avr-stood-still/) - feel free to cut-and-paste anything from there or from the README in that project (https://gitlab.com/snowgoonspub/rust-avr-nightly-builder) to flesh out your blogpost/article.
